### PR TITLE
feat(dependency injection): Add dependency injection for plugins

### LINKED ISSUE #667: Remove side effects from stryker plugins
Right now it is quite difficult to to tryout 1 plugin by linking it in your project (`npm i file:./path/to/plugin`). This is because plugins work based on global side effects.

For example: `foo-transpiler`. The index.js file would look like:

```js
const { TranspilerFactory } = require('stryker-api/transpile');
const FooTranspiler = /*...*/;
TranspilerFactory.instance().register('foo', FooTranspiler);
```

The problem is that the instance of `require('stryker-api/transpile');` is different in the linked package. This is by design (standard nodejs require functionality).

We can solve this issue by not relying on side effects. Instead plugins would export:

```js
const FooTranspiler = /*...*/
exports.FooTranspiler = FooTranspiler;

/* OR in typescript: */
export { FooTranspiler } from './FooTranspiler';
```

@simondel what do you think?

The magic will be the postfix. This could be one of these: `ConfigEditor`, `Transpiler`, `Runner`, `Framework`, `Mutator`.

If a plugin uses one of the factories we should create a deprecation warning. We can remove the factories in the 1.0 release.

UPDATE:

Lot of work is done. Still todo:

- [ ] Replace calls to `getLogger` to DI
- [ ] Deprecate old api (using `process.emitWarning`): https://nodejs.org/dist/latest-v11.x/docs/api/process.html#process_process_emitwarning_warning_type_code_ctor
- [ ] Log deprecated message when loading `config` via DI (using `process.emitWarning`): https://nodejs.org/dist/latest-v11.x/docs/api/process.html#process_process_emitwarning_warning_type_code_ctor

### PR DESCRIPTION
**Pull request by nicojs**
This PR is a preparation of removing the side effects from the way we now load plugins. Instead of registering yourself to the correct factory (for example, by calling `ReporterFactory.instance().register`), it relies on plugins exporting a property called `strykerPlugins`. 

Just to be clear: this PR doesn't fix #667. It only adds the dependency injection mechanism to Stryker. We still need to do PR's for each package (can go relatively fast)

The content of this PR:

1. Add a dependency injection framework called `typed-inject` to the packages folder; A type safe dependency injection framework
1. Add a package `stryker-api/di` which contains dependency injection related interfaces to help plugins know what to (type-safely) inject
1. Add `typed-inject` as a dependency to the core Stryker package and use it. Reporter plugins are loaded via this new mechanism.
1. Alter the `PluginLoader` in order to enable dependency injection in all plugins. It is still backward compatible with the old way of loading plugins.
1. Add a package to contain generic test helpers called `@stryker-mutator/test-helpers`. This package contains a `testInjector` to help inject common stuff into injectables for testability purposes.
1. It updated the `BroadcastReporter` and all build-in reporters to now use the new dependency injection mechanism (as a demo implementation and tryout mostely).
1. It changed the `HtmlReporter` to work with the new way of loading plugins

There is still some work to be done. However, this PR can already be reviewed for the most part. Not changing a lot anymore.

- [ ] Split of the `HtmlReporter` implementation so it can be reviewed in separate PR.
- [ ] Document `typed-inject` in its own readme
- [ ] Add integration tests to `typed-inject` to make sure un-typesafe dependency injection results in compiler errors
- [ ] Remove specific helper methods per plugin: `reporterPlugin`, `configEditorPlugin`, ... in favor of one `plugins` helper method. I think this is possible

### CODE REVIEW

File: packages/typed-inject/README.md [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+
+Scope is also supported here, for more info, see `provideFactory`.
+
+### `Scope`
+
+The `Scope` enum indicates the scope of a provided injectable (class or factory). Possible values: `Scope.Transient` (new injection per resolve) or `Scope.Singleton` (inject once, and reuse values). It generally defaults to `Singleton`. 
+
+### `tokens`
+
+The `tokens` function is a simple helper method that makes sure that an `inject` array is filled with a [tuple type filled with literal strings](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-0.html#rest-parameters-with-tuple-types).
```
Comment (simondel) [inline packages/typed-inject/README.md]: How about calling this what it is? A function to create a tuple filled with literal strings. I think something like `createStringLiteralTuple` would make more sense than `tokens`. What are your thoughts on this?

File: packages/stryker-api/src/plugin/Contexts.ts [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+  options: StrykerOptions;
+  /**
+   * @deprecated This is just here to migrate between old and new plugins. Don't use this! Use `options` instead
+   */
+  config: Config;
+}
+
+export interface PluginContexts {
+  [PluginKind.ConfigEditor]: StrykerContext;
+  [PluginKind.Mutator]: PluginContext;
```
[the remark this answers is not in the document]
Comment (simondel) [reply packages/stryker-api/src/plugin/Contexts.ts]: As discussed: yes.

File: packages/stryker-api/testResources/module/useCore.ts [hunk 3]
```diff
@@ -1,28 +1,46 @@
 import { StrykerOptions, File, Position, Location, Range, LogLevel } from 'stryker-api/core';
 
-const options: StrykerOptions = {};
+const minimalOptions: StrykerOptions = {
```
[the remark this answers is not in the document]
Comment (simondel) [reply packages/stryker-api/testResources/module/useCore.ts]: As discussed: yes.

File: packages/stryker/src/TestFrameworkOrchestrator.ts [hunk 4]
```diff
@@ -1,10 +1,13 @@
 import { TestFrameworkFactory, TestFramework } from 'stryker-api/test_framework';
 import { StrykerOptions } from 'stryker-api/core';
 import { getLogger } from 'stryker-api/logging';
+import { commonTokens } from '@stryker-mutator/util';
+import { tokens } from 'typed-inject';
 
 export default class TestFrameworkOrchestrator {
   private readonly log = getLogger(TestFrameworkOrchestrator.name);
 
+  public static inject = tokens(commonTokens.options);
```
[the remark this answers is not in the document]
Comment (simondel) [reply packages/stryker/src/TestFrameworkOrchestrator.ts]: Fixed.

File: packages/stryker/src/Stryker.ts [hunk 5]
```diff
@@ -1,51 +1,64 @@
...
+    this.applyConfigEditors();
     this.freezeConfig();
-    this.reporter = new ReporterOrchestrator(this.config).createBroadcastReporter();
+    this.injector = rootInjector
+      .provideValue(commonTokens.getLogger, getLogger)
+      .provideFactory(commonTokens.logger, loggerFactory, Scope.Transient)
+      .provideValue(commonTokens.pluginResolver, pluginLoader as PluginResolver)
+      .provideValue(commonTokens.config, this.config)
+      .provideValue(commonTokens.options, this.config as StrykerOptions);
+    this.reporter = this.injector.injectClass(BroadcastReporter);
```
[the remark this answers is not in the document]
Comment (simondel) [reply packages/stryker/src/Stryker.ts]: As discussed: yes.

File: packages/typed-inject/README.md [hunk 6]
```diff
@@ -0,0 +1,264 @@
...
+  .provideClass('httpClient', HttpClient);
+
+const myService = appInjector.injectClass(MyService);
+// Dependencies for MyService validated and injected
+```
+
+In this example: 
+
+* the `logger` is injected into a new instance of `HttpClient` by value.
+* The instance of `HttpClient` and the `logger` is injected into a new instance of `MyService`.
```
[the remark this answers is not in the document]
Comment (nicojs) [reply packages/typed-inject/README.md]: 👍

File: packages/typed-inject/README.md [hunk 7]
```diff
@@ -0,0 +1,264 @@
...
+
+The error messages are a bit cryptic at times, but it sure is better than running into them at runtime.
+
+## ✨ Magic tokens
+
+Any `Injector` instance can always inject the following tokens:
+
+| Token name | Token value | Description |
+| - | - | - | 
+| `TARGET_TOKEN` | `'$target'` | The class or function in which the current values is injected, or `undefined` if resolved directly |
```
[the remark this answers is not in the document]
Comment (nicojs) [reply packages/typed-inject/README.md]: 👍

File: packages/typed-inject/README.md [hunk 8]
```diff
@@ -0,0 +1,264 @@
...
+
+This method creates a new instance of class `injectable` and returns it. 
+When there are any problems in the dependency graph, it gives a compiler error.
+
+```ts
+class Foo {
+    constructor(bar: number) { }
+    static inject = tokens('bar');
+}
+const foo /*: Foo*/ injector.injectClass(Foo);
```
[the remark this answers is not in the document]
Comment (nicojs) [reply packages/typed-inject/README.md]: 👍

File: packages/typed-inject/README.md (hunk 1 above)
Comment (nicojs) [reply packages/typed-inject/README.md]: I think `createStringLiteralTuple` is a far to long name. I understand where you're coming from, but functionally it is used to define _injection tokens_. In the future I'm thinking of expanding the functionality of `typed-inject` to also except Symbols (as well as Strings) using TS 2.7's [`unique symbol`](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-7.html#unique-symbol) type. 

In other words, `tokens` will always be used to define tokens, but might not be used to define solely _string tuples_ in the future. 

Does that make sense?

File: packages/stryker-api/src/plugin/Plugins.ts [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+import { InjectionToken, InjectableClass } from 'typed-inject';
+import { PluginKind } from './PluginKind';
+
+export interface BasePlugin<TPlugin extends PluginKind, R, Tokens extends InjectionToken<PluginContexts[TPlugin]>[]> {
+  readonly kind: TPlugin;
+  readonly name: string;
+  readonly injectable: InjectableClass<PluginContexts[TPlugin], R, Tokens>;
+}
+
+export function reporterPlugin<Tokens extends InjectionToken<PluginContexts[PluginKind.Reporter]>[]>(p: ReporterPlugin<Tokens>) {
```
[the remark this answers is not in the document]
Comment (nicojs) [reply packages/stryker-api/src/plugin/Plugins.ts]: This is out-dated. Use either `declareClassPlugin` or `declareFactoryPlugin`

File: packages/stryker-util/src/tokens.ts [hunk 10]
```diff
@@ -0,0 +1,21 @@
+
+function token<T extends string>(value: T): T {
```
[the remark this answers is not in the document]
Comment (nicojs) [reply packages/stryker-util/src/tokens.ts]: Renamed the method to `stringLiteral` with documentation above. Better? (it was also moved to `stryker-api`).

**Comment by nicojs** [comment]
I've implemented all review comments I think. 

Thanks 👍

File: packages/stryker-api/src/plugin/Plugins.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (simondel) [reply packages/stryker-api/src/plugin/Plugins.ts]: sounds awesome!

File: packages/typed-inject/README.md (hunk 1 above)
Comment (simondel) [reply packages/typed-inject/README.md]: Sounds good!
